### PR TITLE
issue/15 Add setting to choose affiliates list sorting order

### PR DESCRIPTION
**Pull request by dbeja**
Issue #15 

**Housekeeping**

* [x] Issue number has been linked below the description
* [x] Branch name follows the `issue/###` or `release/###` format
* [x] PR title is prefixed with the branch name (issue/### - Title)

**Proposed changes**

* [x] Code has been tested with no errors
* [x] Code meets stated minimum requirements
* [x] Code should (usually) not contain unrelated changes
* [x] Code style (including docs) and organization is generally consistent with the rest of the codebase
* [x] Before and after screenshots have been posted to document visual changes
* [x] Clear steps for testing the PR have been posted (as needed)

**Implemented changes**
* Add setting on add-on settings to choose the sorting order of the affiliates list
* If the user selects "Alphabetical" it will sort ASC by the "Affiliate Display" setting selected

**Before changes**
imagem

**After changes**
imagem

### CODE REVIEW

**Comment by dbeja** [comment]
Do I need to regenerate the language files?

**Comment by DrewAPicture** [comment]
> Do I need to regenerate the language files?

@dbeja As of #25, translations are now handled on .org so no.

File: includes/class-admin.php [hunk 1]
```diff
@@ -55,6 +55,17 @@ public function settings( $fields ) {
 			'std' => 'user_nicename'
 		);
 
+		$fields['checkout_referrals_affiliates_sorting_order'] = array(
+			'name' => __( 'Affiliates Sorting Order', 'affiliatewp-checkout-referrals' ),
+			'desc' => __( 'How the affiliates will be sorted on the checkout dropdown', 'affiliatewp-checkout-referrals' ),
+			'type' => 'radio',
+			'options' => array(
+				'affiliate_id' 	=> 'Affiliate ID',
+				'alphabetical' 	=> 'Alphabetical'
```
Comment (DrewAPicture) [inline includes/class-admin.php:64]:
```suggestion
				'affiliate_id' 	=> __( 'Affiliate ID', 'affiliatewp-checkout-referrals' ),
				'alphabetical' 	=> __( 'Alphabetical', 'affiliatewp-checkout-referrals' ),
```
The option labels need to be translatable.

File: includes/functions.php [hunk 2]
```diff
@@ -38,3 +38,15 @@ function affwp_cr_affiliate_display() {
 function affwp_cr_require_affiliate() {
 	return affiliate_wp()->settings->get( 'checkout_referrals_require_affiliate', false );
 }
+
+/**
+ * Get the affiliates sorting order.
+ * Provides a default for when the settings have not yet been saved.
```
Comment (DrewAPicture) [inline includes/functions.php:44]:
```suggestion
 * Gets the affiliates sorting order.
 *
 * Provides a default for when the settings have not yet been saved.
```
Use a third-person singular verb in the DocBlock summary and add a blank line between the summary and the description.

File: includes/functions.php [hunk 3]
```diff
@@ -38,3 +38,15 @@ function affwp_cr_affiliate_display() {
...
+ *
+ * @since  1.0.9
+ *
+ * @return string affiliates sorting order setting
```
Comment (DrewAPicture) [inline includes/functions.php:48]:
```suggestion
 * @return string Affiliates' sorting order setting.
```
Sentence case for the return description.

File: integrations/class-base.php [hunk 4]
```diff
@@ -49,13 +49,15 @@ public function already_tracking_referral() {
 	 *
 	 * @return array Affiliate IDs and their corresponding user IDs.
 	 */
-	public function get_affiliates() {
+	public function get_affiliates( $args = array() ) {
```
Comment (DrewAPicture) [inline integrations/class-base.php:52]: We're going to need a changelog entry for 1.0.9 for the addition of this optional `$args` parameter as well as an `@param` annotation added to the `get_affiliates()` DocBlock.

File: integrations/class-woocommerce.php [hunk 5]
```diff
@@ -74,12 +74,18 @@ public function affiliate_select_or_input( $checkout ) {
 			return;
 		}
 
-		// get affiliate list
-		$affiliate_list = $this->get_affiliates();
-
 		$description  = affwp_cr_checkout_text();
 		$display      = affwp_cr_affiliate_display();
 		$required     = affwp_cr_require_affiliate();
+		$sorting			= affwp_cr_affiliates_sorting_order();
```
Comment (DrewAPicture) [inline integrations/class-woocommerce.php:80]: Looks like there's some irregular spacing here. Alignment of variables is typically accomplished using spaces, maybe these are tabs?

File: integrations/class-woocommerce.php [hunk 6]
```diff
@@ -74,12 +74,18 @@ public function affiliate_select_or_input( $checkout ) {
...
+
+		// get affiliate list
+		$args = array();
+		if( 'alphabetical' == $sorting ) {
```
Comment (DrewAPicture) [inline integrations/class-woocommerce.php:84]:
```suggestion
		if ( 'alphabetical' === $sorting ) {
```
When in doubt, strict comparison.

**Review by DrewAPicture — CHANGES_REQUESTED** [review]
Just a few docs nitpicks and other standards fixes needed.

**Review by amdrew — CHANGES_REQUESTED** [review]
I'm not seeing the affiliates order changing at checkout when testing with Easy Digital Downloads. 

Looking at the PR code, it seems it may only have been implemented for WooCommerce. We'll need to add support for each integration listed (https://affiliatewp.com/add-ons/official-free/checkout-referrals/): WooCommerce, EDD and Restrict Content Pro)

**Comment by amdrew** [comment]
The ordering (now testing with WooCommerce) doesn't seem to be working as expected. A couple of examples:

## Testing with `Nickname` selected:

Screen Shot 2021-01-11 at 1 11 42 PM

Result:

Screen Shot 2021-01-11 at 1 12 29 PM

Expected result:

The affiliate with `Aaaaa` as a nickname should appear before `Bbbb`

Screen Shot 2021-01-11 at 1 17 44 PM

Screen Shot 2021-01-11 at 1 18 19 PM


---

## Testing with `Display Name`

It's showing "Munro" above `Aaaa`

Screen Shot 2021-01-11 at 1 20 18 PM


Screen Shot 2021-01-11 at 1 16 02 PM

Screen Shot 2021-01-11 at 1 15 47 PM

Screen Shot 2021-01-11 at 1 16 17 PM

File: integrations/class-base.php (hunk 4 above)
Comment (dbeja) [reply integrations/class-base.php]: I don't need this `$args`var anymore, so I reverted back these changes.

**Comment by dbeja** [comment]
> The ordering (now testing with WooCommerce) doesn't seem to be working as expected. A couple of examples:

@amdrew By coincidence, I was using a list of Affiliates on my environment that was equal to the same list sorted alphabetically and so it was always working on my side.
The list of affiliates depends on the AffiliatesWP function `affiliate_wp()->affiliates->get_affiliates( $args )` and this function doesn't support sorting the affiliates by neither `Nicename` nor `Nickname`.
Instead of whitelisting those fields on AffiliatesWP core query, I followed another approach. I get the affiliates list sorted by Affiliate ID (as before) and if the user selected `Alphabetical` I sort the list afterwords on PHP side.

**Comment by dbeja** [comment]
> I'm not seeing the affiliates order changing at checkout when testing with Easy Digital Downloads.
> 
> Looking at the PR code, it seems it may only have been implemented for WooCommerce. We'll need to add support for each integration listed (https://affiliatewp.com/add-ons/official-free/checkout-referrals/): WooCommerce, EDD and Restrict Content Pro)

Missed that. I moved now the function that generates the affiliates list for the "select" to the integrations base class (`class-base.php`) so that it now works on any integration.

**Review by amdrew — APPROVED** [review]
@dbeja Perfecto, all now works as expected 👍

File: integrations/class-base.php [hunk 7]
```diff
@@ -128,13 +163,11 @@ public function show_select_or_input() {
...
+				<?php $affiliates = $this->get_affiliates_select_list( $affiliate_list ); ?>
 
-				<option value="0"><?php _e( 'Select', 'affiliatewp-checkout-referrals' ); ?></option>
-				<?php foreach ( $affiliate_list as $affiliate_id => $user_id ) :
-					$user_info = get_userdata( $user_id );
-				?>
-					<option value="<?php echo $affiliate_id; ?>"><?php echo $user_info->$display; ?></option>
+				<select id="<?php echo $this->context;?>-affiliate" name="<?php echo $this->context;?>_affiliate" class="<?php echo $this->context;?>-select">
+				<?php foreach ( $affiliates as $affiliate_id => $affiliate_display ) : ?>
+					<option value="<?php echo $affiliate_id; ?>"><?php echo $affiliate_display; ?></option>
```
Comment (DrewAPicture) [inline integrations/class-base.php:170]: Need `esc_attr()` on the `$affiliate_id` here.
```suggestion
					<option value="<?php echo esc_attr( $affiliate_id ); ?>"><?php echo $affiliate_display; ?></option>
```

**Review by DrewAPicture — APPROVED** [review]
Looks good.

**Pull request merged**